## Analyzing Used Car Listings on eBay Kleinanzeigen

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The **data dictionary** provided with data is as follows:

- teCrawled - When this ad was first crawled. All field-values are taken from this date.
- me - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

## Project Aim

**The aim of this project is to clean the data and analyze the included used car listings.**

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv(r'C:\Users\Romit\Desktop\Jupyter Notebooks\autos.csv', encoding = "Latin-1")
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [2]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

## Cleaning Data

In [6]:
def clean_col(col):
    col = col.strip()
    col = col.replace('yearOfRegistration','registration_year')
    col = col.replace('monthOfRegistration','registration_month')
    col = col.replace('notRepairedDamage','unrepaired_damage')
    col = col.replace('dateCreated','ad_created')
    col = col.replace('offertype','offer_type')
    col = col.replace('nrofpictures','num_of_pictures')
    col = col.replace('dateCrawled','date_crawled')
    col = col.replace('lastSeen','last_seen')
    col = col.lower()
    return col

modified_cols = []
for c in autos.columns:
    clean_c = clean_col(c)
    modified_cols.append(clean_c)
    
autos.columns = modified_cols
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'postalcode',
       'last_seen'],
      dtype='object')

**We will look for:-**<br> 
Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

The following methods are helpful for exploring the data:-<br>
<code>DataFrame.describe()</code> (with **include='all' to get both categorical and numeric columns**) - <code>Series.value_counts()</code> and <code>Series.head()</code> if any columns need a closer look.

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  num_of_pictures    postalcode  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

From above we observe that:-<br>

- *seller* and *offertype* have mostly same values except one entry
- *nrofpictues* are zero except one entry

*NOTE:* **We have to drop *seller*, *offertype* and *nrofpictures***

- **price and odometer columns** are numeric values stored as text.
- So for each column:
  - Remove any non-numeric characters.
  - Convert the column to a numeric data type.
  - Use DataFrame.rename() to rename the column to odometer_km.

In [8]:
autos = autos.drop(["seller","offer_type","num_of_pictures"], axis = 1)

In [9]:
autos.shape

(50000, 17)

In [10]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [11]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [12]:
autos["odometer"] = autos["odometer"].str.replace('km','').str.replace(",","").astype(int)
autos["price"] = autos["price"].str.replace('$','').str.replace(",","").astype(int)
autos.rename({"odometer" : "odometer_km"}, axis = 1, inplace = True)

In [13]:
autos["odometer_km"].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000], dtype=int64)

In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicletype  registration_year    gearbox  powerps   model  \
0   5000  control         bus               2004    manuell      158  andere   
1   8500  control   limousine               1997  automatik      286     7er   
2   8990     test   limousine               2009    manuell      102    golf   
3   4350  control  kleinwagen               2007  automatik       71  fortwo   
4   1350     test       kombi               2003    manuell        0   focus   

   odometer_km  registration_month fueltype       brand unrepaired_damage  \
0       150000                   3      lpg     peugeot              nein   
1       150000                   6   benzin         bmw              nein   
2        70000                   7   benzin  volkswagen              nein   
3        70000                   6   benzin       smart              nein   
4       150000                   7   benzin        ford              nein   

            ad_created  postalcode            last_seen  
0  2016-03-26 00:00:00       79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00       71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00       35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00       33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00       39218  2016-04-01 14:38:50

## Further Cleaning Data

We'll continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the **odometer_km and price columns.** Here's the steps we'll take:

- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.<br>
- We'll use:<br>
  - <code>Series.unique().shape</code> to see how many unique values
  - <code>Series.describe()</code> to view min/max/median/mean etc
  - <code>Series.value_counts()</code>, with some variations:<br>
    - chained to .head() if there are lots of values.
    - Because <code>Series.value_counts()</code> returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here).<br>
- When removing outliers, we can do df[(df["col"] > x ) & (df["col"] < y )], but it's more readable to use df[df["col"].between(x,y)]

## Unique values in odometer_km and price column

In [15]:
autos["odometer_km"].unique().shape

(13,)

In [16]:
autos["price"].unique().shape

(2357,)

## Viewing the min/max/median/mean values

In [17]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [18]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

## Viewing the values across both the Series

**We start by exploring if the odometer_km column can be cleaned further or not**

**The highest and lowest values in order in odometer_km column**


In [19]:
autos["odometer_km"].value_counts().sort_index(ascending = False).head() # returns the highest values in order from the counts

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

In [20]:
autos["odometer_km"].value_counts().sort_index(ascending = True).head() # returns the lowest values in order from the counts

5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64

All values are rounded, which indicates that the sellers might have to select from pre-set options for this field. There are more high mileage cars.

**We move on to the prices column and explore a bit further to see whether it can be cleaned or not**

In [21]:
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

**The highest and lowest values in order in the prices column**

In [22]:
autos["price"].value_counts().sort_index(ascending = False).head() # returns the highest values in order from the counts

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64

In [23]:
autos["price"].value_counts().sort_index(ascending = True).head() # returns the lowest values in order from the counts

0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64

We observe that most values are rounded, which might indicate that the sellers tend to round-up prices. Additionally, **1421 entries are 0** - given that this is only 2% of the total cars, we might consider removing these. There are a number of listings with **prices less than 50**, and a few listings with **prices more than 1 million.** Given that eBay is an auction site, there could be items where the bid opens at 1. 

*NOTE:*<br>
We will however **remove** items listed **above 350000** since the prices seem to increase steadily after that and it is feasibly not possible to sell cars with such prices on an online website.

In [24]:
autos = autos[autos["price"].between(1,350000)] # removes price below 1 and above 350000
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the Date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself.

    - `date_crawled`: added by the crawler
    - `last_seen`: added by the crawler
    - `ad_created`: from the website
    - `registration_month`: from the website
    - `registration_year`: from the website
    
The **date_crawled**, **last_seen**, and **ad_created** columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively.

In [25]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [26]:
autos['date_crawled'].value_counts().sort_index

<bound method Series.sort_index of 2016-03-19 17:36:18    3
2016-03-29 23:42:13    3
2016-04-02 15:49:30    3
2016-03-23 18:39:34    3
2016-03-22 09:51:06    3
                      ..
2016-04-03 16:39:58    1
2016-03-08 03:59:11    1
2016-03-26 14:40:16    1
2016-03-18 17:44:14    1
2016-03-15 15:49:20    1
Name: date_crawled, Length: 46882, dtype: int64>

**Steps to be followed:**

- We use the workflow we just described to calculate the distribution of values in the date_crawled, ad_created, and last_seen columns (all string columns) as percentages.
  - To include missing values in the distribution and to use percentages instead of counts, chain the       Series.value_counts(**normalize = True, dropna = False**) method.
  - To rank by date in ascending order (earliest to latest), chain the Series.sort_index() method.

In [27]:
# calculates the distribution of values in date_crawled as percentages and includes missing values in the distribution
autos['date_crawled'].value_counts(normalize=True, dropna=False)

2016-03-19 17:36:18    0.000062
2016-03-29 23:42:13    0.000062
2016-04-02 15:49:30    0.000062
2016-03-23 18:39:34    0.000062
2016-03-22 09:51:06    0.000062
                         ...   
2016-04-03 16:39:58    0.000021
2016-03-08 03:59:11    0.000021
2016-03-26 14:40:16    0.000021
2016-03-18 17:44:14    0.000021
2016-03-15 15:49:20    0.000021
Name: date_crawled, Length: 46882, dtype: float64

In [28]:
# now to rank the date_crawled values in ascending order
(autos['date_crawled'].str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_values()
)

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

**Now we repeat the same with ad_created and last_seen columns**

*NOTE:*
Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [29]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

*NOTE:*
There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [30]:
(autos['last_seen'].str[:10]
.value_counts(normalize=True, dropna=False)
.sort_values()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-18    0.007351
2016-03-08    0.007413
2016-03-13    0.008895
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-14    0.012602
2016-03-27    0.015649
2016-03-19    0.015834
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-26    0.016802
2016-03-23    0.018532
2016-03-25    0.019211
2016-03-24    0.019767
2016-03-21    0.020632
2016-03-20    0.020653
2016-03-28    0.020859
2016-03-22    0.021373
2016-03-29    0.022341
2016-04-01    0.022794
2016-03-31    0.023783
2016-03-12    0.023783
2016-04-04    0.024483
2016-03-30    0.024771
2016-04-02    0.024915
2016-04-03    0.025203
2016-03-17    0.028086
2016-04-05    0.124761
2016-04-07    0.131947
2016-04-06    0.221806
Name: last_seen, dtype: float64

*NOTE:*
The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.



In [31]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

**NOTE:**
The year that the car was first registered will likely indicate the age of the car. Looking at the column we realise that the *minimum value is 1000*, way before cars were invented and the *maximum is 9999*, many years into the future.

## Dealing with Incorrect Registration Year Data


Because a car can't be first registered after the listing was seen, any vehicle with a **registration year above 2016 is definitely inaccurate.** Determining the **earliest valid year** is more difficult. Realistically, it could be somewhere in the first **few decades of the 1900s.**

In [51]:
(autos["registration_year"]
.value_counts(normalize=True, dropna=False)
.sort_values()
.head(10)
)

1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1957    0.000043
1951    0.000043
1950    0.000064
1962    0.000086
1956    0.000086
1958    0.000086
Name: registration_year, dtype: float64

In [50]:
autos = autos[autos["registration_year"].between(1950,2016)] # removes registration_years below 1900 and above 2016
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067637
2005    0.062922
1999    0.062086
2004    0.057928
2003    0.057843
2006    0.057221
2001    0.056493
2002    0.053278
1998    0.050642
2007    0.048799
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

## Exploring Price by Brand

In [55]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [56]:
print(autos["brand"].unique().shape)

(40,)


We notice that there are in **total 40 unique brands of cars**.

In [58]:
# These are the top 20 brands of cars in the market listing

autos["brand"].value_counts(normalize=True).head(20)

volkswagen        0.211333
bmw               0.110071
opel              0.107563
mercedes_benz     0.096462
audi              0.086603
ford              0.069866
renault           0.047149
peugeot           0.029854
fiat              0.025653
seat              0.018281
skoda             0.016416
nissan            0.015280
mazda             0.015195
smart             0.014166
citroen           0.014016
toyota            0.012709
hyundai           0.010030
sonstige_autos    0.009644
volvo             0.009151
mini              0.008765
Name: brand, dtype: float64

*NOTE:*
There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [59]:
required_brands = autos["brand"].value_counts(normalize=True)
common_brands = required_brands[required_brands > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


We see that the brands representing **more than 5%** of total listings are as follows:-
- volkswagen
- bmw
- opel
- mercedes_benz
- audi
- ford

In [82]:
# creating an empty dictionary to store the common brands prices
brand_mean_prices = {}

for brand in common_brands:
    brands_only = autos[autos["brand"] == brand]
    mean_price = brands_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices

{'volkswagen': 5398,
 'bmw': 8332,
 'opel': 2976,
 'mercedes_benz': 8625,
 'audi': 9336,
 'ford': 3737}

Of the top 5 brands, there is a distinct price gap:

- Audi being the most expensive followed by BMW and Mercedes Benz.
- Opel is the least expensive and Ford just a bit more expensive than Opel.
- Volkswagen lies somewhere in between - this may explain its popularity among buyers.

## Exploring Mileage

In [83]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5398
bmw                  8332
opel                 2976
mercedes_benz        8625
audi                 9336
ford                 3737

In [84]:
brand_mean_mileage = {}

for brand in common_brands:
    brands_only = autos[autos["brand"] == brand]
    mean_mileage = brands_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [85]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132597
mercedes_benz        130832
opel                 129336
audi                 129157
volkswagen           128710
ford                 124374

In [86]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132597        8332
mercedes_benz        130832        8625
opel                 129336        2976
audi                 129157        9336
volkswagen           128710        5398
ford                 124374        3737